lib/streamlit/elements/lib/mutable_status_container.py

```diff
@@ -0,0 +1,165 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import time
+from types import TracebackType
+from typing import List, Optional, Type, cast
+
+from typing_extensions import Literal, TypeAlias
+
+from streamlit.cursor import Cursor
+from streamlit.delta_generator import DeltaGenerator, _enqueue_message
+from streamlit.proto.Block_pb2 import Block as BlockProto
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+States: TypeAlias = Literal["running", "complete", "error"]
+
+
+class StatusContainer(DeltaGenerator):
+    @staticmethod
+    def _create(
+        parent: DeltaGenerator,
+        label: str,
+        expanded: bool = False,
+        state: States = "running",
+    ) -> StatusContainer:
+        expandable_proto = BlockProto.Expandable()
+        expandable_proto.expanded = expanded
+        expandable_proto.label = label or ""
+
+        if state == "running":
+            expandable_proto.icon = "spinner"
+        elif state == "complete":
+            expandable_proto.icon = "check"
+        elif state == "error":
+            expandable_proto.icon = "error"
+
+        block_proto = BlockProto()
+        block_proto.allow_empty = True
+        block_proto.expandable.CopyFrom(expandable_proto)
+
+        delta_path: List[int] = (
+            parent._active_dg._cursor.delta_path if parent._active_dg._cursor else []
+        )
+
+        status_container = cast(
+            StatusContainer,
+            parent._block(block_proto=block_proto, dg_type=StatusContainer),
+        )
+
+        # Apply initial configuration
+        status_container._delta_path = delta_path
+        status_container._current_proto = block_proto
+        status_container._current_state = state
+
+        # We need to sleep here for a very short time to prevent issues when
+        # the status is updated too quickly. If an .update() directly follows the
+        # the initialization, sometimes only the latest update is applied.
+        # Adding a short timeout here allows the frontend to render the update before.
+        time.sleep(0.05)
+
+        return status_container
+
+    def __init__(
+        self,
+        root_container: int | None,
+        cursor: Cursor | None,
+        parent: DeltaGenerator | None,
+        block_type: str | None,
+    ):
+        super().__init__(root_container, cursor, parent, block_type)
+
+        # Initialized in `_create()`:
+        self._current_proto: BlockProto | None = None
+        self._current_state: States | None = None
+        self._delta_path: List[int] | None = None
+
+    def update(
+        self,
+        *,
+        label: str | None = None,
+        expanded: bool | None = None,
+        state: States | None = None,
+    ) -> None:
+        """Update the status container.
```

Comment:
Should we add something about:

> Only specified arguments are updated. Container contents and unspecified arguments remain unchanged.

Maybe a better way to phrase it. Also does not need to be in the first line one-liner.